packages/cojson-transport-ws/src/index.ts

```diff
@@ -1,108 +1,121 @@
 import { DisconnectedError, Peer, PingTimeoutError, SyncMessage } from "cojson";
-import { Either, Stream, Queue, Effect, Exit } from "effect";
+import { Stream, Queue, Effect, Console } from "effect";
+
+interface WebsocketEvents {
+    close: { code: number; reason: string };
+    message: { data: unknown };
+    open: void;
+}
+interface PingMsg {
+    time: number;
+    dc: string;
+}
 
 interface AnyWebSocket {
-    addEventListener(
-        type: "close",
-        listener: (event: { code: number; reason: string }) => void,
+    addEventListener<K extends keyof WebsocketEvents>(
+        type: K,
+        listener: (event: WebsocketEvents[K]) => void,
     ): void;
-    addEventListener(
-        type: "message",
-        listener: (event: { data: string | unknown }) => void,
+    removeEventListener<K extends keyof WebsocketEvents>(
+        type: K,
+        listener: (event: WebsocketEvents[K]) => void,
     ): void;
-    addEventListener(type: "open", listener: () => void): void;
     close(): void;
     send(data: string): void;
 }
 
+const g: typeof globalThis & {
+    jazzPings?: {
+        received: number;
+        sent: number;
+        dc: string;
+    }[];
+} = globalThis;
+
 export function createWebSocketPeer(options: {
     id: string;
     websocket: AnyWebSocket;
     role: Peer["role"];
 }): Effect.Effect<Peer> {
     return Effect.gen(function* () {
         const ws = options.websocket;
+        const ws_ = ws as unknown as EventTarget;
 
-        const incoming =
-            yield* Queue.unbounded<
-                Either.Either<SyncMessage, DisconnectedError | PingTimeoutError>
-            >();
         const outgoing = yield* Queue.unbounded<SyncMessage>();
 
-        ws.addEventListener("close", (event) => {
-            void Effect.runPromiseExit(
-                Queue.offer(
-                    incoming,
-                    Either.left(
-                        new DisconnectedError(`${event.code}: ${event.reason}`),
-                    ),
-                ),
-            ).then((e) => {
-                if (Exit.isFailure(e) && !Exit.isInterrupted(e)) {
-                    console.warn("Failed closing ws", e);
-                }
-            });
-        });
-
-        let pingTimeout: ReturnType<typeof setTimeout> | undefined;
-
-        ws.addEventListener("message", (event) => {
-            const msg = JSON.parse(event.data as string);
-
-            if (pingTimeout) {
-                clearTimeout(pingTimeout);
-            }
+        const closed = fromEventListener(ws, "close").pipe(
+            Effect.map(
+                (event) =>
+                    new DisconnectedError(`${event.code}: ${event.reason}`),
+            ),
+            Effect.flip,
```

Comment:
```suggestion
            Effect.andThen(
                (event) =>
                    new DisconnectedError(`${event.code}: ${event.reason}`),
            ),
```

a bit shorter don't you think? <picture data-single-emoji=":relieved_galactic:" title=":relieved_galactic:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/galactic_relifed/33851380ad8f06bb.png" alt=":relieved_galactic:" align="absmiddle"></picture> or will it infer the wrong types?

Reply:
`DisconnectError` is not a `YieldableError`, so I have to `Effect.fail` or `Effect.flip`

Reply:
Can we make it a YieldableError?

Reply:
Yeah ofc